### PR TITLE
Explainer Parameters in Request

### PR DESCRIPTION
**Pull request by cliveseldon**
To get effective explanations its important to be able to explore different parameters to the explain process. These should be passed in requests rather than requiring an updated explainer each time.

 * Allows Alibi parameters to be passed as headers in the request with a prefix "Alibi-"

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/520)

### CODE REVIEW

File: python/alibiexplainer/alibiexplainer/anchor_text.py [hunk 1]
```diff
@@ -31,17 +31,20 @@ def __init__(self, predict_fn: Callable, explainer: alibi.explainers.AnchorText,
         self.predict_fn = predict_fn
         self.kwargs = kwargs
         logging.info("Anchor Text args %s", self.kwargs)
+        logging.info("Hey")
```
Comment (rakelkar) [inline python/alibiexplainer/alibiexplainer/anchor_text.py]: Hey hey! :)

File: python/alibiexplainer.Dockerfile [hunk 2]
```diff
@@ -2,7 +2,7 @@ FROM python:3.7
 
 COPY . .
 RUN pip install --upgrade pip && pip install -e ./kfserving
-RUN git clone https://github.com/SeldonIO/alibi.git && \
+RUN git clone -b 'v0.3.2' --single-branch https://github.com/SeldonIO/alibi.git && \
```
Comment (ellis-bigelow) [inline python/alibiexplainer.Dockerfile]: Is alibi not available in pip?

File: python/alibiexplainer/alibiexplainer/anchor_text.py [hunk 3]
```diff
@@ -31,17 +31,20 @@ def __init__(self, predict_fn: Callable, explainer: alibi.explainers.AnchorText,
...
         if explainer is None:
             logging.info("Loading Spacy Language model for %s", spacy_language_model)
             spacy_model(model=spacy_language_model)
             self.nlp = spacy.load(spacy_language_model)
             logging.info("Language model loaded")
         self.anchors_text = explainer
 
-    def explain(self, inputs: List) -> Dict:
+    def explain(self, inputs: List, requestArgs: Dict) -> Dict:
```
Comment (ellis-bigelow) [inline python/alibiexplainer/alibiexplainer/anchor_text.py]: Is it cleaner to just pass in the entire request into explain? That's what the subclassed explain interface does.

**Comment by ellis-bigelow** [comment]
High level comment. Request headers work, but they're not an ideal way for forming these. It seems awkward to try to avoid a standardized protocol by stuffing into headers. Specifically, this doesn't really extend very well into grpc.

We do have a concept that extraneous keys in the request body should be ignored. Do you think we could have some concept of instance_metadata:[] that we pass this as. TF and the other model servers should just ignore the extra field. WDYT?

**Comment by cliveseldon** [comment]
> High level comment. Request headers work, but they're not an ideal way for forming these. It seems awkward to try to avoid a standardized protocol by stuffing into headers. Specifically, this doesn't really extend very well into grpc.
> 
> We do have a concept that extraneous keys in the request body should be ignored. Do you think we could have some concept of instance_metadata:[] that we pass this as. TF and the other model servers should just ignore the extra field. WDYT?

I agree sending the parameters in the payload is preferable. If we can agree on a simple solution I'm happy to replace this PR with one that sends in Request. For example:

```JSON
{
"instance_meta": {
  "explainer": {}
},
 "instances": [1,2,3]
}
```

Just looking for a solution we can use soon so if we are happy with above can redo PR with this?

**Comment by ellis-bigelow** [comment]
> > High level comment. Request headers work, but they're not an ideal way for forming these. It seems awkward to try to avoid a standardized protocol by stuffing into headers. Specifically, this doesn't really extend very well into grpc.
> > We do have a concept that extraneous keys in the request body should be ignored. Do you think we could have some concept of instance_metadata:[] that we pass this as. TF and the other model servers should just ignore the extra field. WDYT?
> 
> I agree sending the parameters in the payload is preferable. If we can agree on a simple solution I'm happy to replace this PR with one that sends in Request. For example:
> 
> ```json
> {
> "instance_meta": {
>   "explainer": {}
> },
>  "instances": [1,2,3]
> }
> ```
> 
> Just looking for a solution we can use soon so if we are happy with above can redo PR with this?

This looks fine to me. I don't think we should treat this interface as anything beyond Alpha quality, but SGTM to get customer feedback. I'd prefer instance_metadata (fully spelled out)

**Comment by cliveseldon** [comment]
Will open new PR for this

**Pull request closed without merging**